extmod: factor mbedtls config to common location, and enable elliptic curves

dpgeorge opened the pull request:
This PR factors all mbedtls config to `extmod/mbedtls/mbedtls_config_common.h`, and then enables EC-DH and EC-DSA on all ports that use it (which is rp2, mimxrt, stm32, unix).

ECHS/ECDSA is necessary to access sites that only support these protocols (eg `calendarific.com`).
    
The rp2 port already has ECDH enabled, so this just adds ECDSA.  The other ports now gain both ECDH and ECDSA.  The code size increase is:
    
- rp2 (PICO_W): +2916 bytes flash, +24 bytes BSS
- stm32 (PYBD_SF6): +20480 bytes flash, +32 bytes data, +48 bytes BSS
- mimxrt (TEENSY41): +20708 bytes flash, +32 bytes data, +48 bytes BSS
- unix (standard x86-64): +39344 executable, +1744 bytes data, +96 BSS

dpgeorge committed:
extmod/mbedtls: Add common configuration file, and use it in all ports.

This is a no-op change.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/mbedtls: Enable MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_KEY_EXCHANGE.

This was already enabled on all ports except mimxrt.  Now it's enabled on
all of them.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/mbedtls: Enable elliptic curve DH and DSA cryptography.

This is necessary to access sites that only support these protocols.

The rp2 port already has ECDH enabled, so this just adds ECDSA there.  The
other ports now gain both ECDH and ECDSA.  The code size increase is:

- rp2 (PICO_W): +2916 bytes flash, +24 bytes BSS
- stm32 (PYBD_SF6): +20480 bytes flash, +32 bytes data, +48 bytes BSS
- mimxrt (TEENSY41): +20708 bytes flash, +32 bytes data, +48 bytes BSS
- unix (standard x86-64): +39344 executable, +1744 bytes data, +96 BSS

This is obviously a large increase in code size.  But there doesn't seem to
be any other option because without elliptic curve cryptography devices are
partially cut off from the internet.  For use cases that require small
firmware size, they'll need to build custom firmware with a custom mbedtls
config.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
This is obviously a large increase in code size (for the builds that can afford it).  But I don't see any other option.  Without elliptic curve cryptography, MicroPython devices are partially cut off from the internet.

For users that need to reduce code size they'll need to build custom firmware with a custom mbedtls config.

jimmo commented on ports/rp2/mbedtls/mbedtls_config.h:
```diff
@@ -26,90 +26,23 @@
-#define MBEDTLS_ECP_C
-#define MBEDTLS_ENTROPY_C
-#define MBEDTLS_ERROR_C
 #define MBEDTLS_GCM_C
```
Can we remove this exception from the rp2 config and enable for all?

ucryptolib doesn't enable GCM, so the only reason for enabling it would be for SSL. If we think that sites use it (it is a valid TLS configuration), then we should enable it for all ports?

dpgeorge replied:
I would have to check exactly what it does (and also code size increase).

Carglglz commented:
> This is obviously a large increase in code size (for the builds that can afford it). But I don't see any other option. Without elliptic curve cryptography, MicroPython devices are partially cut off from the internet.

Definitely, for reference [Server_Side_TLS](https://wiki.mozilla.org/Security/Server_Side_TLS) recommendations from Mozilla.
(from python ssl docs) 

> For users that need to reduce code size they'll need to build custom firmware with a custom mbedtls config.

In case anyone needs this: [Reducing Mbed TLS memory and storage footprint](https://mbed-tls.readthedocs.io/en/latest/kb/how-to/reduce-polarssl-memory-and-storage-footprint/)

dpgeorge commented on ports/rp2/mbedtls/mbedtls_config.h:
```diff
@@ -26,90 +26,23 @@
-#define MBEDTLS_ECP_DP_BP384R1_ENABLED
-#define MBEDTLS_ECP_DP_BP512R1_ENABLED
-#define MBEDTLS_ECP_DP_CURVE25519_ENABLED
 #define MBEDTLS_ECP_NIST_OPTIM
```
It costs quite a bit of flash, maybe 6k.  On unix it's not needed because it's already fast there.  On rp2 it's probably worth it because the rp2 CPU is not fast.  I was going to do some perf tests on stm32 with this option enabled/disabled to see if it made a difference, but that takes time.

dpgeorge commented:
Thanks @Carglglz , both of those links are helpful!

dpgeorge commented:
The `ssl_data.py` multitest now fails, due to this PR.  It's because the server wants to use elliptic curves but the supplied key/cert in `ssl_data.py` is not "good enough" anymore.

So this PR is on hold until we resolve that issue.

Carglglz commented:
> The `ssl_data.py` multitest now fails, due to this PR. It's because the server wants to use elliptic curves but the supplied key/cert in `ssl_data.py` is not "good enough" anymore.
> 
> So this PR is on hold until we resolve that issue.

I can reproduce this, `#define MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED` option is what makes it fail. Not sure why, but `ssl_cert_rsa.py` test passed, maybe is time to deprecate`ssl_data.py` and add `ssl_cert_ec.py`?   (I do have this test already and it passed too)

dpgeorge commented:
> maybe is time to deprecate`ssl_data.py` and add `ssl_cert_ec.py`?

But deprecating it will be a breaking change for users who have built a web server using older encryption standards.  Would be best to not break `ssl_data.py`.

jimmo commented:
> > The `ssl_data.py` multitest now fails, due to this PR. It's because the server wants to use elliptic curves but the supplied key/cert in `ssl_data.py` is not "good enough" anymore.
> > So this PR is on hold until we resolve that issue.
> 
> I can reproduce this, `#define MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED` option is what makes it fail. Not sure why, but `ssl_cert_rsa.py` test passed, maybe is time to deprecate`ssl_data.py` and add `ssl_cert_ec.py`? (I do have this test already and it passed too)

The problem here is that it's negotiating the highest supported cipher by both ends (now ECDHE), but then the server is presenting an RSA cert.

I think what we need to do is support the `ciphers` kwarg to `wrap_socket` (this is also in CPython, as well as on `SSLContext.set_ciphers`), which will take a comma separated list of mbedtls cipher names to turn into a list of integers to update `o->conf` with after https://github.com/micropython/micropython/blob/master/extmod/modussl_mbedtls.c#L188

Could use mbedtls_ssl_ciphersuite_from_string etc...

We'll have to make the default something that maintains backwards compatibility (as Damien just commented above).

Carglglz commented:
> > maybe is time to deprecate`ssl_data.py` and add `ssl_cert_ec.py`?
> 
> But deprecating it will be a breaking change for users who have built a web server using older encryption standards. Would be best to not break `ssl_data.py`.

I agree with not making breaking changes, I meant to change key/cert in `ssl_data.py` to use EC key/cert. 
Right now `ssl_data.py` key is RSA 512 bits, which is definitely not a real use case (at least it needs 2048 bits as per recommendations above).
```
DBG:../../lib/mbedtls/library/ssl_srv.c:0816: cert. version     : 1
...
DBG:../../lib/mbedtls/library/ssl_srv.c:0754: signed using      : RSA with SHA1

DBG:../../lib/mbedtls/library/ssl_srv.c:0754: RSA key size      : 512 bits
```

> The problem here is that it's negotiating the highest supported cipher by both ends (now ECDHE), but then the server is presenting an RSA cert.

I did enabled `#define MBEDTLS_DEBUG_C` to see what's really going on and looks like cipher suite is 

```
DBG:../../lib/mbedtls/library/ssl_srv.c:2024: selected ciphersuite: TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA384
```
And failure happens here
```
+DBG:../../lib/mbedtls/library/ssl_srv.c:3274: mbedtls_pk_sign() returned -16512 (-0x4080)
+
+DBG:../../lib/mbedtls/library/ssl_tls.c:8216: <= handshake
+
+DBG:../../lib/mbedtls/library/ssl_tls.c:9060: => free
+
+DBG:../../lib/mbedtls/library/ssl_tls.c:9125: <= free
+
+Traceback (most recent call last):
+  File "<stdin>", line 121, in <module>
+  File "<stdin>", line 51, in instance0
+  File "ssl.py", line 1, in wrap_socket
+OSError: (-16512, 'MBEDTLS_ERR_RSA_BAD_INPUT_DATA')
```

Running `ssl_cert_rsa.py` it uses same cipher suite 
```
DBG:../../lib/mbedtls/library/ssl_srv.c:2024: selected ciphersuite: TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA384
```
The only difference is in key/cert 
```
DBG:../../lib/mbedtls/library/ssl_tls.c:5489: cert. version     : 3
...
DBG:../../lib/mbedtls/library/ssl_tls.c:5489: signed using      : RSA with SHA-256

DBG:../../lib/mbedtls/library/ssl_tls.c:5489: RSA key size      : 4096 bits
```
And it does pass the test. So it looks to me like this cipher suite `TLS-ECDHE-RSA-WITH-AES-256-CBC-SHA384` does not allow RSA 512 bits key size.

If `MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED` is disabled, the cipher suite selected
```
+DBG:../../lib/mbedtls/library/ssl_srv.c:2024: selected ciphersuite: TLS-RSA-WITH-AES-256-CBC-SHA256
```
allows this key size and the test runs without failures.

Carglglz commented:
@jimmo 
> I think what we need to do is support the ciphers kwarg to wrap_socket (this is also in CPython, as well as on SSLContext.set_ciphers), 

I've already done this (kind of) in #8968 (Only one cipher suite at the moment, but I could change it to make it accept a list)

I've added `ciphers` kwarg to `wrap_socket` and set to `TLS-RSA-WITH-AES-256-CBC-SHA256` cipher suite. The test passed as expected, however I still would update the RSA key to at least 2048 bits and add a `ssl_cert_ec.py` test...

Carglglz commented:
@dpgeorge 
If you don't want to modify `ssl_data.py` by any means disabling `#define MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED` for unix port is the only option I see right now. It won't be a breaking change since there would be plenty of cipher suites (RSA and ECDSA based) that server and client can agree while doing the handshake...

Also to avoid confusion`#define MBEDTLS_ECP_DP_CURVE25519_ENABLED` can be disabled since it is not used for TLS right now (It's not implemented in MbedTLS, there are some PRs and it is supposed to be on the roadmap but no ETA as of now...)

dpgeorge commented:
> If you don't want to modify `ssl_data.py` by any means

I would like to keep this test as-is.  It provides a good baseline test with a simple (albeit old/weak) key/cert pair.

> disabling `#define MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED` for unix port is the only option I see right now.

Yes, that could be a solution to get this PR merged.  The rp2 port has that enabled already, and it can just stay enabled in that port only, no other ports will enable it (yet).

> so to avoid confusion`#define MBEDTLS_ECP_DP_CURVE25519_ENABLED` can be disabled

OK.

dpgeorge commented:
> Also to avoid confusion`#define MBEDTLS_ECP_DP_CURVE25519_ENABLED` can be disabled since it is not used for TLS right now (It's not implemented in MbedTLS, there are some PRs and it is supposed to be on the roadmap but no ETA as of now...

@Carglglz can you provide more info about this?  It looks like there's code in mbedtls to implement this curve, and disabling it reduces code size on the unix port by around 4.5k.  So it must do something...?

Carglglz commented:
@dpgeorge 
> @Carglglz can you provide more info about this? It looks like there's code in mbedtls to implement this curve, and disabling it reduces code size on the unix port by around 4.5k. So it must do something...?

Yes, full info here [curve25519 support](https://mbed-tls.readthedocs.io/en/latest/kb/cryptography/elliptic-curve-performance-nist-vs-brainpool/?highlight=ed25519#curve25519-support)

The key part:

> Unfortunately, they use slightly different data structures and representations than the other curves, so they haven’t been ported yet to TLS and PKIX in Mbed TLS. We do support basic Curve25519 arithmetic though.

> So it must do something...?

Yes, there is a sample `ecdh_curve25519.c` program in `mbedtls/programs/pkey` where it performs a  `ECDHE` key exchange, but that's it, unfortunately no TLS implementation as of now (see [EdDSA software implementation](https://github.com/orgs/Mbed-TLS/projects/1#column-17950143) project of MbedTLS github)

Also based on performance 
```
# NIST 
ECDSA-secp521r1          :    1093 sign/s
ECDSA-secp384r1          :    1556 sign/s
ECDSA-secp256r1          :    2121 sign/s
ECDSA-secp224r1          :    3103 sign/s
ECDSA-secp192r1          :    4107 sign/s
ECDSA-secp521r1          :     299 verify/s
ECDSA-secp384r1          :     431 verify/s
ECDSA-secp256r1          :     612 verify/s
ECDSA-secp224r1          :     935 verify/s
ECDSA-secp192r1          :    1316 verify/s

# BRAINPOOL
ECDSA-brainpoolP512r1    :     163 sign/s
ECDSA-brainpoolP384r1    :     361 sign/s
ECDSA-brainpoolP256r1    :     603 sign/s
ECDSA-brainpoolP512r1    :      37 verify/s
ECDSA-brainpoolP384r1    :      81 verify/s
ECDSA-brainpoolP256r1    :     156 verify/s
```

I would disable Brainpool curves, and even maybe a subset of NIST could be good enough e.g. only these three.
```
#define MBEDTLS_ECP_DP_SECP256R1_ENABLED
#define MBEDTLS_ECP_DP_SECP384R1_ENABLED
#define MBEDTLS_ECP_DP_SECP521R1_ENABLED
```
I'm pretty sure these ones are widely supported.

dpgeorge committed:
extmod/mbedtls: Remove MBEDTLS_ECP_DP_CURVE25519_ENABLED config.

Curve25519 arithmetic is supported in mbedtls, but it's not used for TLS.
So there's no need to have this option enabled.

Reduces rp2 PICO_W firmware by 2440 bytes.

Thanks to @Carglglz for the information.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/mbedtls: Remove brainpool curves from config.

They are much slower than NIST (SECP) curves and shouldn't be needed.

Reduces rp2 PICO_W firmware by 1328 bytes.

Thanks to @Carglglz for the information.

Signed-off-by: Damien George <damien@micropython.org>
